### PR TITLE
Implement `filter_rules` for Profiles

### PR DESCRIPTION
**Pull request by JAORMX**
This implements a mechanism to filter rules for a profile based on
attributes of the rule object.

This allows us to do stuff like this:

```
filter_rules: 'platform == "ocp4-node"'
```

... Since the filter takes the Python syntax, more complex rules can be
created. e.g. one is able to use all boolean operators from the
language.

The variables available are exactly the ones available in the `Rule`
object from *build_yaml.py*.

Signed-off-by: Juan Antonio Osorio Robles <jaosorior@redhat.com>

### CODE REVIEW

File: products/ocp4/profiles/pci-dss-node.profile [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        - mrogers950
+
+reference: https://www.pcisecuritystandards.org/documents/PCI_DSS_v3-2-1.pdf
+
+title: 'PCI-DSS v3.2.1 Control Baseline for Red Hat OpenShift Container Platform 4'
+
+description: |-
+    Ensures PCI-DSS v3.2.1 security configuration settings are applied.
+
+filter_rules: 'platform == "ocp4-node"'
```
Comment (yuumasato) [inline products/ocp4/profiles/pci-dss-node.profile:18]: It feels strange to have this separated from the control selection item (i.e. `pcidss_ocp4:all:base`).

Do we also want/need the `filter_rules` to work independently of control based profiles?

File: ssg/build_yaml.py [hunk 2]
```diff
@@ -1032,6 +1042,16 @@ def __str__(self):
         return self.id_
 
 
+def rule_filter_from_def(filterdef):
+    if filterdef is None or filterdef == "":
+        return None
+    def filterfunc(rule):
+        # Remove globals for security and only expose
+        # variables relevant to the rule
+        return eval(filterdef, {}, rule.__dict__)
```
Comment (yuumasato) [inline ssg/build_yaml.py:1051]: Neat!

File: products/ocp4/profiles/pci-dss-node.profile (hunk 1 above)
Comment (yuumasato) [reply products/ocp4/profiles/pci-dss-node.profile:18]: Reading the description again there is no mention of the controls file. I read the PR with a different context in mind, :stuck_out_tongue:

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now a5e4be3b471e5b99918f6bc5e6e668bbfe410188

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now f75ff0507e90eb0503c1bad95b357a0ea8b6d159

**Comment by JAORMX** [comment]
Pull-request updated, HEAD is now c7f387b56681f445e946c542806747e658de941a

**Comment by ggbecker** [comment]
@JAORMX please check https://github.com/JAORMX/content/pull/4 it fixes the problems with the controls unit testing.

**Comment by JAORMX** [comment]
/retest

IdP deployment error (took too long)

**Review by yuumasato — CHANGES_REQUESTED** [review]
I like the filtering mechanism. But what is its expected scope? Is it expected to work with profiles defined without control files?

I tried restricting RHEL-8 `ospp` and `anssi_bp_28_minimal` profiles with `filter_rules: 'platform == "machine"'`, and it seems to have worked only with the `anssi` profile.

I think the behavior should be extended to all profiles, or the syntax mechanism integrated into the "control selection mechanism":

`     - pcidss_ocp4:all:base['platform == "ocp4-node"']`

Edit: testing with  `'platform != "machine"'` the difference is more evident.

File: products/ocp4/profiles/pci-dss-node.profile (hunk 1 above)
Comment (matejak) [reply products/ocp4/profiles/pci-dss-node.profile:18]: We have discussed the question of interface - should this filtering clause be here, in the profile file, or should it be closer to the controls manipulation, because that's what we want - to influence what rules we get from those controls?

I suggest that we keep it as it is for now - we can easily change the interface to something else, as the code that asks rules about their platform would stay, and only the place it is invoked from would change.
By looking at `pcidss_ocp4:all:base`, it is clear that the control syntax has been abused enough, and `pcidss_ocp4:all:base:platform == something` could be the solution for now, but wouldn't benefit the project in the long-term. So let's come with a better-looking and more robust and intuitive instantiation syntax, and only then let's use it.

File: ssg/controls.py [hunk 3]
```diff
@@ -114,18 +114,19 @@ def from_control_dict(cls, control_dict, env_yaml=None, default_level=["default"
                 varname, value = item.split("=", 1)
                 control.variables[varname] = value
             else:
+                rule_yaml = get_rule_path_by_id(content_dir, item)
```
Comment (matejak) [inline ssg/controls.py:117]: This is OK for now, but generally rules loaded this way won't posses inherited group platforms, and in some cases they may not be valid YAMLs due to jinja2 constructs. However, this problem can be solved later.

File: ssg/build_yaml.py [hunk 4]
```diff
@@ -1498,6 +1531,9 @@ def to_file(self, file_name):
         tree = ET.ElementTree(root)
         tree.write(file_name)
 
+    def __str__(self):
+        return self.id_
```
Comment (matejak) [inline ssg/build_yaml.py:1535]: Do we need a string representation of a rule? A human-readable representation would be rather its title, and a reliable representation is basically its yaml.

**Comment by matejak** [comment]
> I like the filtering mechanism. But what is its expected scope? Is it expected to work with profiles defined without control files?
> 

I can imagine that we could get clean container profiles "for free", which could make sense.

> I tried restricting RHEL-8 `ospp` and `anssi_bp_28_minimal` profiles with `filter_rules: 'platform == "machine"'`, and it seems to have worked only with the `anssi` profile.

It could be caused by wrong detection of platforms because of group inheritance. I would expect `"machine" in / not in platforms` to do the right thing, but that's not possible ATM. But I don't see this as a blocker to the PR, as the target product OCP4 doesn't suffer from those problems, and the behavior can be fixed for the rest by making the build process more sane.

> 
> I think the behavior should be extended to all profiles, or the syntax mechanism integrated into the "control selection mechanism":

See also my reply https://github.com/ComplianceAsCode/content/pull/7588#discussion_r713162890 - I suggest not to abuse the control line any more, but I see your point and I can't say that I would disagree.

@JAORMX What was the reason that controls got the list of data structures rather than rule IDs?

File: products/ocp4/profiles/pci-dss-node.profile (hunk 1 above)
Comment (yuumasato) [reply products/ocp4/profiles/pci-dss-node.profile:18]: > We have discussed the question of interface - should this filtering clause be here, in the profile file, or should it be closer to the controls manipulation, because that's what we want - to influence what rules we get from those controls?

TL;DR; I tend to think that that the filtering clause should be in the profile.
--
The control file keeps a set of controls organized according to the policy it is tracking, it organizes the controls into policy levels and maps rules to them. It can be seen as an organized "rule gatherer".

Now the pool of rules gathered by the control file need to make it into profiles.
And, although the profiles can be aligned to the policy (ANSSI), they can be organized in an unrelated way (PCI-DSS for OCP4). What I mean by that is that the separation into `master` and `node` is an OCP4 construct, the policy doesn't predict the product being "divided" into two parts.

So I think the filter should be defined in the profile to avoid adding product specific constructs in the control file.

> By looking at `pcidss_ocp4:all:base`, it is clear that the control syntax has been abused enough, and `pcidss_ocp4:all:base:platform == something` could be the solution for now, but wouldn't benefit the project in the long-term - that would be the third selection option added to the one-line syntax, as first we select `all` controls, we then restrict those with `level == base`, and then we have this platform filter. It is likely that another selection requirements are yet to be discovered. So let's come with a better-looking and more robust and intuitive instantiation syntax, and only then let's use it.

I agree that the control syntax is being pushed beyond what it was designed for, which, as far as I remember, was to be backward compatible with the profile syntax.
Maybe consider backward compatibility breaking changes?

**Comment by JAORMX** [comment]
> > I like the filtering mechanism. But what is its expected scope? Is it expected to work with profiles defined without control files?
> 
> I can imagine that we could get clean container profiles "for free", which could make sense.
> 
> > I tried restricting RHEL-8 `ospp` and `anssi_bp_28_minimal` profiles with `filter_rules: 'platform == "machine"'`, and it seems to have worked only with the `anssi` profile.
> 
> It could be caused by wrong detection of platforms because of group inheritance. I would expect `"machine" in / not in platforms` to do the right thing, but that's not possible ATM. But I don't see this as a blocker to the PR, as the target product OCP4 doesn't suffer from those problems, and the behavior can be fixed for the rest by making the build process more sane.

OK, should we then address more functionality in a subsequent PR? We could document this until it's ready then. But this would be a first iteration.

> 
> > I think the behavior should be extended to all profiles, or the syntax mechanism integrated into the "control selection mechanism":
> 
> See also my reply [#7588 (comment)](https://github.com/ComplianceAsCode/content/pull/7588#discussion_r713162890) - I suggest not to abuse the control line any more, but I see your point and I can't say that I would disagree.
> 
> @JAORMX What was the reason that controls got the list of data structures rather than rule IDs?

It allowed me to easily traverse the rules and get all the data from them, such as the platform. I can't do that just with the rule IDs.

File: ssg/build_yaml.py (hunk 4 above)
Comment (JAORMX) [reply ssg/build_yaml.py:1535]: @matejak this is needed for utils/controls-template.html

File: ssg/build_yaml.py (hunk 2 above)
Comment (matejak) [reply ssg/build_yaml.py:1052]:
```suggestion
        return eval(filterdef, {"__builtins__": None}, rule.__dict__)
```

According to https://stackoverflow.com/a/9558001/592892 this can be improved, but eval is still dangerous, although that modification restricts the attack surface.

**Review by yuumasato — APPROVED** [review]
The intended use with control files work, but it will need to be iterated to work with profiles defined without controls.
